### PR TITLE
Iss32

### PR DESCRIPTION
**Pull request by ArtPoon**

### CODE REVIEW

File: openrdp/scripts/bootscan.py [hunk 1]
```diff
@@ -109,48 +116,86 @@ def find_potential_events(self, pair1, pair2):
...
-        dists = []
+        num_arrays = self.num_replicates
+        array_shape = ((self.align.shape[0] - 1) * (self.align.shape[0]) //2,)
+        dists = np.empty((num_arrays,) + array_shape, dtype=np.float16)
         np.random.seed(self.random_seed)
         random.seed(self.random_seed)
         for rep in range(self.num_replicates):
             # Shuffle columns with replacement
             rep_window = window[:, np.random.randint(0, window.shape[1], window.shape[1])]
-            dist_mat = squareform(pdist(rep_window, jc_distance))
```
Comment (ArtPoon) [inline openrdp/scripts/bootscan.py:118]: This is part of a merge conflict.  I think we can discard `squareform` in `dev`, which is converting the more compact pdist vector into the full square matrix.

File: openrdp/scripts/bootscan.py [hunk 2]
```diff
@@ -109,48 +116,86 @@ def find_potential_events(self, pair1, pair2):
...
-            all_dists = p.map(self.scan, range(0, align.shape[1], self.step_size))
 
-        return all_dists
+        # Remove previous .h5 files
+        h5_files = glob.glob('*.h5')
+        for file in h5_files:
+            if os.path.exists(file):
+                os.remove(file)
+
+        with multiprocessing.Pool(self.np) as p:
```
Comment (ArtPoon) [inline openrdp/scripts/bootscan.py:151]: next part of merge conflict - I don't think there are any issues here, this is @GopiGugan 's method of writing data to filesystem before running multiple threads.

File: openrdp/scripts/bootscan.py [hunk 3]
```diff
@@ -109,48 +116,86 @@ def find_potential_events(self, pair1, pair2):
...
+        triplet = Triplet(self.align, self.seq_names, trp)
+
+        print("Scanning triplet {} / {}".format(i, self.total_triplet_combinations))
 
         # Look at boostrap support for sequence pairs
-        ab_support = []
-        bc_support = []
-        ac_support = []
-        for dists in self.dists:
+        ab_support = [0] * (self.align.shape[1] // self.step_size)
```
Comment (ArtPoon) [inline openrdp/scripts/bootscan.py:170]: Third part of merge conflict - pre-allocating these vectors is probably a good idea.

File: openrdp/scripts/bootscan.py [hunk 4]
```diff
@@ -215,9 +260,17 @@ def execute(self, triplet):
...
-                            self.raw_results.append((rec_name, parents, *event, val))
+                            raw_results.append((rec_name, parents, *event, val))
+
+        return raw_results
+
+    def execute_all(self, total_combinations, seq_names):
+        self.seq_names = seq_names
+        self.total_triplet_combinations = total_combinations
+        with multiprocessing.Pool(self.np) as p:
+            results = p.map(self.execute, enumerate(generate_triplets(self.align)))
```
Comment (ArtPoon) [inline openrdp/scripts/bootscan.py:271]: Last part of merge conflict.  @dpannguyen moved a conditional statement out of the loop over `trp_names`.  @GopiGugan is this compatible with your changes?

Comment (GopiGugan) [reply openrdp/scripts/bootscan.py:273]: Yes, that doesn't conflict with the changes I made

**Review by GopiGugan — APPROVED** [review]
OpenRDP runs as expected after merging `dev` into `iss32`. Still need to verify the correctness of the output

**Comment by ArtPoon** [comment]
@GopiGugan have you had a chance to verify outputs?

**Pull request merged**